src/expr/src/interpret.rs

```diff
@@ -0,0 +1,1180 @@
+/// (However, if either `f(b)` and `f(c)` error or return null, we don't assume anything about `f(a)`.
+///
+/// This should likely be moved to a function annotation in the future.
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+enum Monotonic {
+    /// Monotonic in this argument.
+    Yes,
+    /// We haven't classified this one yet. (Conservatively treated as "no", but a distinct
+    /// enum so it's easy to tell whether someone's thought about a particular function yet.)
+    /// This should likely be removed when this becomes a function annotation.
+    Maybe,
+    /// We don't know any useful properties of this function.
+    No,
+}
+
+fn unary_monotonic(func: &UnaryFunc) -> Monotonic {
+    use crate::func::impls as funcs;
+    use Monotonic::*;
+    use UnaryFunc::*;
+    match func {
+        // Casts are generally monotonic.
+        CastUint64ToNumeric(_)
+        | CastNumericToUint64(_)
+        | CastNumericToMzTimestamp(_)
+        | CastUint64ToMzTimestamp(_)
+        | CastTimestampToMzTimestamp(_)
+        | CastTimestampTzToMzTimestamp(_) => Yes,
+        // Including JSON casts.
+        CastJsonbToNumeric(_)
+        | CastJsonbToBool(_)
+        | CastJsonbToString(_)
+        | CastJsonbToInt64(_)
+        | CastJsonbToFloat64(_) => Yes,
+
+        // Extracting the "most significant bits" of the a timestamp is monotonic.
+        ExtractTimestamp(funcs::ExtractTimestamp(units))
+        | ExtractTimestampTz(funcs::ExtractTimestampTz(units)) => match units {
+            DateTimeUnits::Epoch
+            | DateTimeUnits::Millennium
+            | DateTimeUnits::Century
+            | DateTimeUnits::Decade
+            | DateTimeUnits::Year => Yes,
+            _ => No,
+        },
+        DateTruncTimestamp(_) | DateTruncTimestampTz(_) => Yes,
+        // Negation is monotonically decreasing, but that's fine.
+        NegInt64(_) | NegNumeric(_) | Not(_) => Yes,
+        // Trivially monotonic, since all non-null values map to `false`.
+        IsNull(_) => Yes,
+        AbsInt64(_) => No,
+        _ => Maybe,
+    }
+}
+
+fn binary_monotonic(func: &BinaryFunc) -> (Monotonic, Monotonic) {
+    use BinaryFunc::*;
+    use Monotonic::*;
+
+    match func {
+        AddInt64 | MulInt64 | AddNumeric | MulNumeric | SubInt64 | SubNumeric => (Yes, Yes),
+        AddInterval | SubInterval => (Yes, Yes),
+        AddTimestampInterval
+        | AddTimestampTzInterval
+        | SubTimestampInterval
+        | SubTimestampTzInterval => (Yes, Yes),
+        // Monotonic in the left argument, but the right hand side has a discontinuity.
+        // Could be treated as monotonic if we also captured the valid domain of the function args.
+        DivInt64 | DivNumeric => (Yes, No),
+        Lt | Lte | Gt | Gte => (Yes, Yes),
+        _ => (Maybe, Maybe),
+    }
+}
+
+fn variadic_monotonic(func: &VariadicFunc) -> Monotonic {
+    use Monotonic::*;
+    use VariadicFunc::*;
+    match func {
+        And | Or => Yes,
+        _ => Maybe,
```

Comment:
Same. `Coalesce` is definitely monotone; other functions may be worth a quick look.